File: wrappers/pyagrum/pyLibs/causal/_doorCriteria.py
############################################################################
#   This file is part of the aGrUM/pyAgrum library.                        #
#                                                                          #
#   Copyright (c) 2005-2025 by                                             #
#       - Pierre-Henri WUILLEMIN(_at_LIP6)                                 #
#       - Christophe GONZALES(_at_AMU)                                     #
#                                                                          #
#   The aGrUM/pyAgrum library is free software; you can redistribute it    #
#   and/or modify it under the terms of either :                           #
#                                                                          #
#    - the GNU Lesser General Public License as published by               #
#      the Free Software Foundation, either version 3 of the License,      #
#      or (at your option) any later version,                              #
#    - the MIT license (MIT),                                              #
#    - or both in dual license, as here.                                   #
#                                                                          #
#   (see https://agrum.gitlab.io/articles/dual-licenses-lgplv3mit.html)    #
#                                                                          #
#   This aGrUM/pyAgrum library is distributed in the hope that it will be  #
#   useful, but WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,          #
#   INCLUDING BUT NOT LIMITED TO THE WARRANTIES MERCHANTABILITY or FITNESS #
#   FOR A PARTICULAR PURPOSE  AND NONINFRINGEMENT. IN NO EVENT SHALL THE   #
#   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER #
#   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,        #
#   ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR  #
#   OTHER DEALINGS IN THE SOFTWARE.                                        #
#                                                                          #
#   See the GNU Lesser General Public License (LICENSE.LGPL) and the MIT   #
#   licence (LICENSE.MIT) for more details.                                #
#                                                                          #
#   Contact  : info_at_agrum_dot_org                                       #
#   homepage : http://agrum.gitlab.io                                      #
#   gitlab   : https://gitlab.com/agrumery/agrum                           #
#                                                                          #
############################################################################

"""
This files define some routines for front and back door
"""

from typing import Iterator, Optional
import itertools as it

import pyagrum

from pyagrum.causal._types import NameSet, NodeId, NodeSet, NodeList
from pyagrum.causal._dSeparation import isParent, dSep_reduce, descendants, isDSep_parents

# pylint: disable=unused-import
import pyagrum.causal  # for annotations


def backdoor_path(bn: "pyagrum.BayesNet", x: str, y: str, zset: NameSet = None) -> bool:
  """
  Predicate on the existence of an open back door path from ``x`` to ``y``,
  conditioning on the set of variables ``zset``

  Parameters
  ----------
  bn: pyagrum.BayesNez
      the DAG model
  x: str
      name of source node
  y: str
      name of destination node
  zset: Set[str]
      names of conditioning nodes

  Returns
  -------
  bool
      True if such backdoor exists
  """
  if zset is None:
    zset = set()
  return not isDSep_parents(bn, x, y, zset)


def exists_unblocked_directed_path(bn: "pyagrum.BayesNet", x: str, y: str, zset: NameSet = None) -> bool:
  """
  Predicate on the existence of a directed path from ``x`` to ``y`` in the Bayesian network ``bn``
  not blocked by nodes of ``zset``

  Parameters
  ----------
  bn: pyagrum.BayesNez
      the DAG model
  x: str
      name of source node
  y: str
      name of destination node
  zset: Set[str]
      names of conditioning nodes

  Returns
  -------
  bool
  """
  if zset is None:
    zset = set()
  if y in bn.children(x):
    return True

  for c in bn.children(x):
    if (c not in zset) and exists_unblocked_directed_path(bn, c, y, zset):
      return True
  return False


def is_frontdoor(bn: "pyagrum.BayesNet", x: str, y: str, zset: NameSet) -> bool:
  """
  Tests whether or not ``zset`` satisifies the front door criterion for ``x`` and ``y``, in the Bayesian network ``bn``

  Parameters
  ----------
  bn: pyagrum.BayesNez
      the DAG model
  x: str
      name of source node
  y: str
      name of destination node
  zset: Set[str]
      names of conditioning nodes

  Returns
  -------
  bool
  """
  if exists_unblocked_directed_path(bn, x, y, zset):
    return False

  if len(zset & backdoor_reach(bn, x)) != 0:
    return False

  bn_reduit = dSep_reduce(bn, zset | {x, y})
  for i in zset:
    if backdoor_path(bn_reduit, i, y, {x}):
      return False

  return True


def is_backdoor(bn: "pyagrum.BayesNet", x: str, y: str, zset: NameSet) -> bool:
  """
  Tests whether or not ``zset`` satisifies the back door criterion for ``x`` and ``y``, in the Bayesian network ``bn``

  Parameters
  ----------
  bn: pyagrum.BayesNez
      the DAG model
  x: str
      name of source node
  y: str
      name of destination node
  zset: Set[str]
      names of conditioning nodes

  Returns
  -------
  bool
  """

  desc_x = descendants(bn, x)
  if len(desc_x & zset) != 0:
    return False

  return isDSep_parents(bn, x, y, zset)


def backdoor_reach(bn: "pyagrum.BayesNet", a: NodeId) -> NodeSet:
  """
  Returns the set of nodes that can be reached through a backdoor path from ``a`` in the graph ``bn``

  Parameters
  ----------
  bn: pyagrum.BayesNet
    the model
  a: int
    the backdoor-ed node

  Returns
  -------
  Set[int]
    set of nodes
  """

  def inner_br(bn: "pyagrum.BayesNet", x: NodeId, pht: bool, reach0: NodeSet, reach1: NodeSet):
    for c in bn.children(x):
      if c not in reach0 and c not in reach1:
        reach1.add(c)
        inner_br(bn, c, True, reach0, reach1)

    if not pht:
      for p in bn.parents(x):
        if p not in reach0:
          reach0.add(p)
          inner_br(bn, p, False, reach0, reach1)

  r = {a}
  r.union(bn.parents(a))
  l = {a}
  for pa in bn.parents(a):
    inner_br(bn, pa, False, r, l)
  s = r | l
  if a in s:
    s.remove(a)
  return s


def nodes_on_dipath(bn: "pyagrum.BayesNet", x: NodeId, y: NodeId) -> Optional[NodeSet]:
  """
  Returns the set of nodes through which there is a directed path from `x` to `y` in the graph `bn`

  Parameters
  ----------
  bn: pyagrum.BayesNet
  x  int
  y: int

  Returns
  -------
  Set[int] (maybe None)
  """

  def inner_nod(g: "pyagrum.BayesNet", a: NodeId, b: NodeId) -> Optional[NodeSet]:
    if b == a:
      return set()

    inners = {a}
    children = g.children(a)
    if len(children) == 0:
      return None

    found = False
    for c in children:
      s = inner_nod(g, c, b)
      if s is not None:
        found = True
        inners |= s
    if found:
      return inners
    return None

  r = inner_nod(bn, x, y)
  if r:
    r.remove(x)

  return r


def backdoor_generator(bn: "pyagrum.BayesNet", cause: NodeId, effect: NodeId, not_bd: NodeSet = None):
  """
  Generates backdoor sets for the pair of nodes `(cause, effect)` in the graph `bn` excluding the nodes in the set
  `not_bd` (optional)

  Parameters
  ----------
  bn: pyagrum.BayesNet
  cause: int
  effect: int
  not_bd: Set[int] default=None

  Yields
  -------
  List[int]
    the different backdoors
  """
  if len(bn.parents(cause)) == 0:  # no parent of cause, no backdoor
    return
  if isParent(effect, cause, bn):  # causalDagFromBN(bn)):
    return

  if not_bd is None:
    not_bd = set()

  # simplifying the graph
  interest = {cause, effect}
  G = dSep_reduce(bn, interest)

  # removing the non connected in G without descendants
  # GG is a trash graph just to find the disjointed nodes in G
  GG = pyagrum.DiGraph(G)
  for i in descendants(bn, cause, set()):
    GG.eraseNode(i)

  # we only keep interesting connex components
  for nodes in GG.connectedComponents().values():
    if nodes.isdisjoint(interest):
      for n in nodes:
        G.eraseNode(n)

  possible = set(G.nodes()) - (descendants(bn, cause, set()) | interest | not_bd)
  if len(possible) == 0:
    return

  backdoors = set()

  for i in range(len(possible)):
    for subset in it.combinations(possible, i + 1):
      sub = frozenset(subset)
      worth_testing = True
      for s in backdoors:
        if s <= sub:
          worth_testing = False
      if worth_testing and isDSep_parents(G, {cause}, {effect}, sub):
        backdoors.add(sub)
        yield list(subset)


def frontdoor_generator(bn: "pyagrum.BayesNet", x: NodeId, y: NodeId, not_fd: NodeSet = None):
  """
  Generates frontdoor sets for the pair of nodes `(x, y)` in the graph `bn` excluding the nodes in the set
  `not_fd` (optional)

  Parameters
  ----------
  bn: pyagrum.BayesNet
  x: int
  y: int
  not_fd: Set[int] default=None

  Yields
  -------
  List[int]
    the different frontdoors
  """
  if isParent(x, y, bn):
    return

  if not_fd is None:
    not_fd = set()

  possible = nodes_on_dipath(bn, x, y)

  nodiPath = False
  if possible is None:
    nodiPath = True
    possible=None
    for _,cc in bn.connectedComponents().items():
      if x in cc and y in cc:
        possible=cc
    if possible is None:
      return
    possible -= {x, y}

  possible -= backdoor_reach(bn, x)
  possible -= not_fd
  impossible = set()
  g = dSep_reduce(bn, {x, y} | possible)
  for z in possible:
    if backdoor_path(g, {z}, {y}, {x}):
      impossible.add(z)
  possible -= impossible
  frontdoors = set()

  if nodiPath:
    for s in possible:
      yield [s]
    return

  for i in range(len(possible)):
    for subset in it.combinations(possible, i + 1):
      sub = frozenset(subset)
      worth_testing = True
      for s in frontdoors:
        if s <= sub:
          worth_testing = False
      if worth_testing and not exists_unblocked_directed_path(bn, x, y, sub):
        frontdoors.add(sub)
        yield list(subset)

File: src/agrum/CM/tools/doorCriteria.cpp
/****************************************************************************
 *   This file is part of the aGrUM/pyAgrum library.                        *
 *                                                                          *
 *   Copyright (c) 2005-2025 by                                             *
 *       - Pierre-Henri WUILLEMIN(_at_LIP6)                                 *
 *       - Christophe GONZALES(_at_AMU)                                     *
 *                                                                          *
 *   The aGrUM/pyAgrum library is free software; you can redistribute it    *
 *   and/or modify it under the terms of either :                           *
 *                                                                          *
 *    - the GNU Lesser General Public License as published by               *
 *      the Free Software Foundation, either version 3 of the License,      *
 *      or (at your option) any later version,                              *
 *    - the MIT license (MIT),                                              *
 *    - or both in dual license, as here.                                   *
 *                                                                          *
 *   (see https://agrum.gitlab.io/articles/dual-licenses-lgplv3mit.html)    *
 *                                                                          *
 *   This aGrUM/pyAgrum library is distributed in the hope that it will be  *
 *   useful, but WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,          *
 *   INCLUDING BUT NOT LIMITED TO THE WARRANTIES MERCHANTABILITY or FITNESS *
 *   FOR A PARTICULAR PURPOSE  AND NONINFRINGEMENT. IN NO EVENT SHALL THE   *
 *   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER *
 *   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,        *
 *   ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR  *
 *   OTHER DEALINGS IN THE SOFTWARE.                                        *
 *                                                                          *
 *   See LICENCES for more details.                                         *
 *                                                                          *
 *   SPDX-FileCopyrightText: Copyright 2005-2025                            *
 *       - Pierre-Henri WUILLEMIN(_at_LIP6)                                 *
 *       - Christophe GONZALES(_at_AMU)                                     *
 *   SPDX-License-Identifier: LGPL-3.0-or-later OR MIT                      *
 *                                                                          *
 *   Contact  : info_at_agrum_dot_org                                       *
 *   homepage : http://agrum.gitlab.io                                      *
 *   gitlab   : https://gitlab.com/agrumery/agrum                           *
 *                                                                          *
 ****************************************************************************/
/**
 * @file
 * @brief Door criteria (backdoor/frontdoor) utilities bound to a single DAG.
 */

#include <agrum/CM/tools/doorCriteria.h>

#include <algorithm>
#include <vector>

#include <agrum/CM/tools/separation.h>

namespace gum {

namespace {
// Convert NodeSet to sorted std::vector<NodeId> for deterministic ops.
static std::vector<NodeId> _sortedVec(const NodeSet& s) {
  std::vector<NodeId> v;
  v.reserve(s.size());
  for (auto n : s) v.push_back(n);
  std::sort(v.begin(), v.end());
  return v;
}

// Deterministic lexicographic comparator for NodeSet via their sorted vectors.
static bool _lexLess(const NodeSet& a, const NodeSet& b) {
  auto va = _sortedVec(a);
  auto vb = _sortedVec(b);
  return va < vb;
}
} // namespace

/* ========================= Backdoor / Frontdoor ========================= */

bool DoorCriteria::satisfiesBackdoorCriterion(NodeId X, NodeId Y, const NodeSet& Z) const {
  // Rule-out illegal contents up front:
  // - Z must not contain X (conditioning on X is not a valid adjustment set)
  // - Z must not contain Y (keep parity with enumeration/universe)
  if (Z.contains(X) || Z.contains(Y)) return false;

  // - Z must not contain any descendant of X
  NodeSet descX = _dag.descendants(X);
  for (auto dx : descX) {
    if (Z.contains(dx)) return false;
  }

  // Backdoor blocking condition in G_{underline X}
  NodeSet Xs; Xs.insert(X);
  NodeSet Ys; Ys.insert(Y);
  return Separation::isBackdoorSeparated(_dag, Xs, Ys, Z);
}

bool DoorCriteria::satisfiesFrontdoorCriterion(NodeId X, NodeId Y, const NodeSet& Z) const {
  if (Z.empty()) return false;

  // (FD-1) Z must intercept all directed X->...->Y paths
  if (existsUnblockedDirectedPath(X, Y, Z)) return false;

  // (FD-2) Z ∩ backdoor_reach(X) = ∅
  NodeSet br = backdoorReach(X);
  for (auto z : Z) if (br.contains(z)) return false;

  // (FD-3) check on a d-separation reduced graph with interest = Z ∪ {X,Y}
  NodeSet Xset; Xset.insert(X);
  NodeSet Yset; Yset.insert(Y);
  // interest = X ∪ Y ∪ Z  -> pass Xset, Yset, Z as-is
  DAG reduced = Separation::reduceForDSeparation(_dag, Xset, Yset, Z);

  NodeSet condX; condX.insert(X);
  for (auto z : Z) {
    NodeSet Zs; Zs.insert(z);
    // backdoor_path(reduced, z, y, {x}) == !isBackdoorSeparated(reduced, {z}, {y}, {x})
    if (!Separation::isBackdoorSeparated(reduced, Zs, Yset, condX)) return false;
  }
  return true;
}




DoorCriteria::NodeSetVec
DoorCriteria::enumerateBackdoorSets(NodeId X, NodeId Y, const EnumerationOptions& opts, bool stopAtFirst) const {
  NodeSetVec out;

  // Theoretical variant: if X has no parents, {} is a valid backdoor set.
  if (_dag.parents(X).empty()) { out.emplace_back(); return out; }

  // If Y is a parent of X, there is an open backdoor X<-Y that {} cannot block.
  if (_dag.existsArc(Y, X)) return out;

  // Reduce around interest = {X, Y}; evidence = ∅
  NodeSet Xset; Xset.insert(X);
  NodeSet Yset; Yset.insert(Y);
  NodeSet Zempty;
  DAG G = Separation::reduceForDSeparation(_dag, Xset, Yset, Zempty);

  // Candidate pool = nodes(G) \ (Desc(X) ∪ {X,Y} ∪ opts.exclude)
  NodeSet descX = _dag.descendants(X); // Python uses descendants(bn, cause)
  NodeSet possible;
  for (auto n : G.nodes()) {
    if (n == X || n == Y) continue;
    if (opts.excluded_nodes.contains(n)) continue;
    if (descX.contains(n)) continue;
    possible.insert(n);
  }
  if (possible.empty()) return out;

  auto cand = _sortedVec(possible);
  const std::size_t N    = cand.size();
  const std::size_t Kmax = (opts.max_cardinality == 0) ? N : std::min(opts.max_cardinality, N);

  // Only prune supersets in minimal_only mode
  NodeSetVec chosen;
  const bool prune_supersets = opts.only_minimal;

  NodeSet Xs; Xs.insert(X);
  NodeSet Ys; Ys.insert(Y);

  // pyagrum tries k >= 1 (never yields {} except our explicit no-parent case above)
  for (std::size_t k = 1; k <= Kmax; ++k) {
    std::vector<bool> pick(N, false);
    std::fill(pick.begin(), pick.begin() + k, true);
    do {
      NodeSet Z;
      for (std::size_t i = 0; i < N; ++i) if (pick[i]) Z.insert(cand[i]);

      if (prune_supersets) {
        bool worth = true;
        for (const auto& s : chosen) {
          bool subset = true;
          for (auto u : s) if (!Z.contains(u)) { subset = false; break; }
          if (subset) { worth = false; break; }
        }
        if (!worth) continue;
      }

      if (Separation::isBackdoorSeparated(G, Xs, Ys, Z)) {
        if (opts.only_minimal) {
          chosen.push_back(Z);
          out.push_back(Z);
        } else {
          out.push_back(Z); // keep non-minimal supersets too (e.g., {U,W})
        }
        if (stopAtFirst) return out;
      }

    } while (std::prev_permutation(pick.begin(), pick.end()));
  }

  // Deterministic order + dedup
  std::sort(out.begin(), out.end(), _lexLess);
  out.erase(std::unique(out.begin(), out.end(),
                        [](const NodeSet& a, const NodeSet& b){
                          return _sortedVec(a) == _sortedVec(b);
                        }),
            out.end());
  return out;
}



DoorCriteria::NodeSetVec
DoorCriteria::enumerateBackdoorSets(NodeId X, NodeId Y, bool stopAtFirst) const {
  EnumerationOptions opts;
  return enumerateBackdoorSets(X, Y, opts, stopAtFirst);
}

// Backward compatibility: default stopAtFirst = false
DoorCriteria::NodeSetVec
DoorCriteria::enumerateBackdoorSets(NodeId X, NodeId Y) const {
  return enumerateBackdoorSets(X, Y, false);
}


DoorCriteria::NodeSetVec
DoorCriteria::enumerateFrontdoorSets(NodeId X, NodeId Y, const EnumerationOptions& opts, bool stopAtFirst) const {
  NodeSetVec out;

  // pyagrum early exit: if X is a parent of Y, yield nothing
  for (auto p : _dag.parents(Y)) if (p == X) return out;

  // Candidate pool
  NodeSet possible = nodesOnDirectedPaths(X, Y);
  bool noDiPath = false;

  if (possible.empty()) {
    // No directed path: use weakly connected component containing both X and Y (undirected BFS)
    NodeSet cc;
    {
      std::vector<NodeId> st; st.push_back(X);
      cc.insert(X);
      while (!st.empty()) {
        NodeId u = st.back(); st.pop_back();
        for (auto v : _dag.parents(u)) if (!cc.contains(v)) { cc.insert(v); st.push_back(v); }
        for (auto v : _dag.children(u)) if (!cc.contains(v)) { cc.insert(v); st.push_back(v); }
      }
    }
    if (!cc.contains(Y)) return out;
    noDiPath = true;
    possible = cc;
    possible.erase(X);
    possible.erase(Y);
  } else {
    possible.erase(X);
    possible.erase(Y);
  }

  // Prune by backdoor reach of X and user exclusions
  NodeSet br = backdoorReach(X);
  for (auto n : br)         possible.erase(n);
  for (auto n : opts.excluded_nodes) possible.erase(n);

  // Remove "impossible" z with a backdoor to Y given X, on a reduced graph:
  // interest = {X, Y} ∪ possible; evidence = ∅ (parity with pyagrum)
  NodeSet Xset; Xset.insert(X);
  NodeSet Yset = possible; Yset.insert(Y);
  NodeSet Zempty;

  DAG g = Separation::reduceForDSeparation(_dag, Xset, Yset, Zempty);

  NodeSet impossible;
  NodeSet condX; condX.insert(X);
  for (auto z : possible) {
    NodeSet Zs; Zs.insert(z);
    if (!Separation::isBackdoorSeparated(g, Zs, NodeSet{Y}, condX)) {
      impossible.insert(z);
    }
  }
  for (auto z : impossible) possible.erase(z);

  if (possible.empty()) return out;

  auto cand = _sortedVec(possible);

  if (noDiPath) {
    // pyagrum: yield each singleton
    for (auto n : cand) {
      NodeSet Z; Z.insert(n); out.push_back(Z);
      if (stopAtFirst) return out;
    }
  } else {
    const std::size_t N    = cand.size();
    const std::size_t Kmax = (opts.max_cardinality == 0) ? N : std::min(opts.max_cardinality, N);
    for (std::size_t k = 1; k <= Kmax; ++k) {
      std::vector<bool> pick(N, false);
      std::fill(pick.begin(), pick.begin() + k, true);
      do {
        NodeSet Z;
        for (std::size_t i = 0; i < N; ++i) if (pick[i]) Z.insert(cand[i]);
        // (FD-1): must block all directed X->Y paths
        if (!existsUnblockedDirectedPath(X, Y, Z)) {
          if (!opts.only_minimal || _isMinimalFrontdoorAdjustment(X, Y, Z)) {
            out.push_back(Z);
            if (stopAtFirst) return out;
          }
        }
      } while (std::prev_permutation(pick.begin(), pick.end()));
    }
  }

  std::sort(out.begin(), out.end(), _lexLess);
  out.erase(std::unique(out.begin(), out.end(),
                        [](const NodeSet& a, const NodeSet& b){ return _sortedVec(a) == _sortedVec(b); }),
            out.end());
  return out;
}



DoorCriteria::NodeSetVec
DoorCriteria::enumerateFrontdoorSets(NodeId X, NodeId Y, bool stopAtFirst) const {
  EnumerationOptions opts;
  return enumerateFrontdoorSets(X, Y, opts, stopAtFirst);
}

// Backward compatibility: default stopAtFirst = false
DoorCriteria::NodeSetVec
DoorCriteria::enumerateFrontdoorSets(NodeId X, NodeId Y) const {
  return enumerateFrontdoorSets(X, Y, false);
}

/* ============================== Utilities =============================== */

bool DoorCriteria::existsUnblockedDirectedPath(NodeId X, NodeId Y, const NodeSet& Z) const {
  // DFS over directed edges; internal nodes blocked by membership in Z.
  if (_dag.existsArc(X, Y)) return true;

  for (auto c : _dag.children(X)) {
    if (!Z.contains(c) && existsUnblockedDirectedPath(c, Y, Z)) return true;
  }
  return false;
}

NodeSet DoorCriteria::nodesOnDirectedPaths(NodeId X, NodeId Y) const {
  NodeSet f = _dag.descendants(X); // forward reach
  NodeSet r = _dag.ancestors(Y);  // reverse reach

  NodeSet res;
  // A node is on some X->...->Y path iff reachable from X and can reach Y.
  for (auto n : f) if (r.contains(n)) res.insert(n);

  // Include endpoints if connected (i.e., Y is forward-reachable from X).
  if (f.contains(Y)) { res.insert(X); res.insert(Y); }

  return res;
}

// pyagrum.backdoor_reach inner function:
// inner_br(bn, x, pht, reach0, reach1)
// - always explore children into reach1
// - explore parents into reach0 only if pht == false
static void _inner_br(const DAG& dag, NodeId x, bool pht,
                      NodeSet& reach0, NodeSet& reach1) {
  // children phase
  for (auto c : dag.children(x)) {
    if (!reach0.contains(c) && !reach1.contains(c)) {
      reach1.insert(c);
      _inner_br(dag, c, true, reach0, reach1);  // after child step, forbid next parent step
    }
  }
  // parents phase (only if we did not just take a parent step)
  if (!pht) {
    for (auto p : dag.parents(x)) {
      if (!reach0.contains(p)) {
        reach0.insert(p);
        _inner_br(dag, p, false, reach0, reach1);
      }
    }
  }
}



NodeSet DoorCriteria::backdoorReach(NodeId a) const {
  NodeSet reach0; // via a parent step
  NodeSet reach1; // via a child step

  // Seed with 'a' to avoid parent->a->child zig-zag artifacts
  reach0.insert(a);
  reach1.insert(a);

  // parents of X are themselves in backdoor reach (X <- parent)
  for (auto pa : _dag.parents(a)) reach0.insert(pa);

  // Start the alternating traversal from each parent
  for (auto pa : _dag.parents(a)) {
    _inner_br(_dag, pa, /*pht=*/false, reach0, reach1);
  }

  NodeSet s = reach0;
  for (auto n : reach1) s.insert(n);
  s.erase(a);
  return s;
}


bool DoorCriteria::hasBackdoorPath(NodeId X, NodeId Y, const NodeSet& Z) const {
  NodeSet Xs; Xs.insert(X);
  NodeSet Ys; Ys.insert(Y);
  return !Separation::isBackdoorSeparated(_dag, Xs, Ys, Z);
}

/* ============================= Minimality =============================== */

bool DoorCriteria::_isMinimalBackdoorAdjustment(NodeId X, NodeId Y, const NodeSet& Z) const {
  for (auto z : Z) {
    NodeSet sub = Z; sub.erase(z);
    if (satisfiesBackdoorCriterion(X, Y, sub)) return false;
  }
  return true;
}

bool DoorCriteria::_isMinimalFrontdoorAdjustment(NodeId X, NodeId Y, const NodeSet& Z) const {
  if (Z.empty()) return false; // frontdoor sets must be non-empty
  for (auto z : Z) {
    NodeSet sub = Z; sub.erase(z);
    if (satisfiesFrontdoorCriterion(X, Y, sub)) return false;
  }
  return true;
}

} // namespace gum
